cranelift/wasm/src/environ/zkasm.rs

```diff
@@ -0,0 +1,953 @@
+        module: &'data str,
+        field: &'data str,
+    ) -> WasmResult<()> {
+        self.info.memories.push(Exportable::new(memory));
+        self.info
+            .imported_memories
+            .push((String::from(module), String::from(field)));
+        Ok(())
+    }
+
+    fn declare_data_initialization(
+        &mut self,
+        _memory_index: MemoryIndex,
+        _base: Option<GlobalIndex>,
+        _offset: u64,
+        _data: &'data [u8],
+    ) -> WasmResult<()> {
+        // We do nothing
+        Ok(())
+    }
+
+    fn declare_func_export(&mut self, func_index: FuncIndex, name: &'data str) -> WasmResult<()> {
+        self.info.functions[func_index]
+            .export_names
+            .push(String::from(name));
+        Ok(())
+    }
+
+    fn declare_table_export(
+        &mut self,
+        table_index: TableIndex,
+        name: &'data str,
+    ) -> WasmResult<()> {
+        self.info.tables[table_index]
+            .export_names
+            .push(String::from(name));
+        Ok(())
+    }
+
+    fn declare_memory_export(
+        &mut self,
+        memory_index: MemoryIndex,
+        name: &'data str,
+    ) -> WasmResult<()> {
+        self.info.memories[memory_index]
+            .export_names
+            .push(String::from(name));
+        Ok(())
+    }
+
+    fn declare_global_export(
+        &mut self,
+        global_index: GlobalIndex,
+        name: &'data str,
+    ) -> WasmResult<()> {
+        self.info.globals[global_index]
+            .export_names
+            .push(String::from(name));
+        Ok(())
+    }
+
+    fn declare_start_func(&mut self, func_index: FuncIndex) -> WasmResult<()> {
+        debug_assert!(self.info.start_func.is_none());
+        self.info.start_func = Some(func_index);
+        Ok(())
+    }
+
+    fn define_function_body(
+        &mut self,
+        mut validator: FuncValidator<ValidatorResources>,
+        body: FunctionBody<'data>,
+    ) -> WasmResult<()> {
+        self.func_bytecode_sizes
+            .push(body.get_binary_reader().bytes_remaining());
+        let func = {
+            let mut func_environ =
+                ZkasmFuncEnvironment::new(&self.info, self.expected_reachability.clone());
+            let func_index =
+                FuncIndex::new(self.get_num_func_imports() + self.info.function_bodies.len());
+
+            let sig = func_environ.vmctx_sig(self.get_func_type(func_index));
+            let mut func =
+                ir::Function::with_name_signature(UserFuncName::user(0, func_index.as_u32()), sig);
+
+            self.trans
+                .translate_body(&mut validator, body, &mut func, &mut func_environ)?;
+            func
+        };
+        self.info.function_bodies.push(func);
+        Ok(())
+    }
+
+    fn declare_module_name(&mut self, name: &'data str) {
+        self.module_name = Some(String::from(name));
+    }
+
+    fn declare_func_name(&mut self, func_index: FuncIndex, name: &'data str) {
```

Comment:
I imagine this is where you'd obtain "useful" symbol names for your functions? Though, hm, those aren’t guaranteed to remain present in wasm modules (there is an optional custom name section – not sure if cranelift's translator interprets that…)

Reply:
That's a good point! We should investigate whether we can use those for better label names in functions.